src/backend/distributed/executor/multi_executor.c

```diff
@@ -617,12 +617,24 @@ IsLocalReferenceTableJoinPlan(PlannedStmt *plan)
 		return false;
 	}
 
-	foreach(oidCell, plan->relationOids)
+	/*
+	 * plan->rtable contains the flattened RTE lists of the plan tree, which
+	 * includes rtes in subqueries, CTEs, ...
+	 *
+	 * It doesn't contain optimized away table accesses (due to join optimization),
+	 * which is fine for our purpose.
+	 */
+	foreach(rangeTableCell, plan->rtable)
 	{
-		Oid relationId = lfirst_oid(oidCell);
+		RangeTblEntry *rangeTableEntry = (RangeTblEntry *) lfirst(rangeTableCell);
 		bool onlySearchPath = false;
 
-		if (RelationIsAKnownShard(relationId, onlySearchPath))
+		if (rangeTableEntry->rtekind != RTE_RELATION)
+		{
+			continue;
```

Comment:
Since views are RTE_RELATION they are also treated as local tables.

```sql
CREATE VIEW v AS SELECT * FROM ref WHERE a = 1;
BEGIN;
SELECT * FROM v JOIN ref ON (v.a = ref.a);
ERROR:  cannot join local tables and reference tables in a transaction block, udf block, or distributed CTE subquery
```

Reply:
Fixed + tests. Also added tests and comments for the function join.